custom/icds_reports/views.py

```diff
@@ -1718,3 +1721,63 @@ def get(self, request, *args, **kwargs):
             return export_response(sync.get_file_from_blobdb(), 'csv', zip_name)
         except NotFound:
             raise Http404
+
+
+class DishaAPIView(View):
+
+    def message(self, message_name):
+        state_names = ", ".join(self.valid_state_names)
+        error_messages = {
+            "missing_date": "Please specify valid month and year",
+            "invalid_month": "Please specify a month that's older than a month and 5 days",
+            "invalid_state": "Please specify one of {} as state_name".format(state_names),
+            "data_is_unavailable": "No data is available for this month"
+        }
+        return {"message": error_messages[message_name]}
+
+    @method_decorator(api_auth)
+    @json_error
+    def get(self, request, *args, **kwargs):
+        # Todo; add disha API permission
+        try:
+            month = int(request.GET.get('month'))
+            year = int(request.GET.get('year'))
+        except (ValueError, TypeError):
+            return JsonResponse(self.messages('missing_date'), status=400)
+
+        # Can return only one month old data if today is after 5th, otherwise
+        #   can return two month's old data
+        query_month = date(year, month, 1)
+        today = date.today()
+        current_month = today - relativedelta(months=1) if today.day <= 5 else today
+        if query_month > current_month:
+            return JsonResponse(self.messages('invalid_month'), status=400)
+
+        state_name = self.request.GET.get('state_name')
+        if state_name not in self.valid_state_names:
+            return JsonResponse(self.messages('invalid_state'), status=400)
+
+        return self._get_json_data(query_month, state_name)
+
+    @property
+    @quickcache([])
+    def valid_state_names(self):
+        return AwcLocationMonths.objects.values_list('state_name', flat=True)
+
+    def _get_json_data(self, query_month, state_name):
+        # Todo; cache monthly payloads
+        columns = [field.name for field in DishaIndicatorView._meta.fields]
+        columns.remove("month")
+        data = DishaIndicatorView.objects.filter(
+            month=query_month,
+            state_name__iexact=state_name
+        ).values_list(*columns)
+        if not len(data):
+            return JsonResponse(self.messages('data_is_unavailable'), status=404)
```

Comment:
404 seems like the wrong response since it signals an error condition, but no data does not seem like an error condition. What do you think of returning the JSON data with an empty list of rows?

Reply:
Sure